### PR TITLE
 Sdam diagnostics

### PR DESCRIPTION
**Pull request by p-mongo**
This primarily adds the summary method to sdam-related objects. It is marked experimental because it may change as we work on sdam.

As https://jira.mongodb.org/browse/RUBY-1443 involves file and class splits I would like to get this diff into master now.

### CODE REVIEW

**Comment by p-mongo** [comment]
https://gist.github.com/p-mongo/69b1b02ab9a09b752be804490c54b830

File: lib/mongo/error/no_server_available.rb [hunk 1]
```diff
@@ -27,9 +27,11 @@ class NoServerAvailable < Error
       #
       # @param [ Hash ] server_selector The server preference that could not be
       #   satisfied.
+      # @param [ Cluster ] cluster The cluster that server selection was
+      #   performed on. (added in 2.7.0)
       #
       # @since 2.0.0
-      def initialize(server_selector)
+      def initialize(server_selector, cluster=nil)
```
Comment (saghm) [inline lib/mongo/error/no_server_available.rb]: Ditto question on #1011; why is the cluster parameter here?

Comment (p-mongo) [reply lib/mongo/error/no_server_available.rb]: Must've gotten killed off.

**Comment by p-mongo** [comment]
Diagnostics when trying to connect with RS name mismatch: https://gist.github.com/p-mongo/bdcb99c45bf2025332d668a4fd9f4370

**Pull request merged**